mabwiser/popularity.py

```diff
@@ -33,9 +33,17 @@ def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.n
 
     def predict(self, contexts: np.ndarray = None) -> Arm:
 
-        # Select an arm randomized by expectation probability
+        # Return the first arm with maximum expectation
+        return argmax(self.predict_expectations())
+
+    def predict_expectations(self, contexts: np.ndarray = None) -> Dict[Arm, Num]:
+
+        # Return a random value between 0 and 1 for each arm that is "proportional" to the
```

Comment:
Huh, I think I understand why it isn't straightforward now. Does this create the same effect as random sampling based on the expectations? As in, if we continuously sampled from the previous implementation vs the current implementation, would we get a similar distribution of items sampled? As long as the _distribution_ of items as number of samples goes to infinity is the same as `self.arm_to_expectations`, I don't have a problem with the implementation.

Reply:
A bit of a sidenote, but I'm also wondering if we're cornering ourselves by calling this function `predict_expectations`. The "expectations" in this case _are_ static regardless of the context (as evidenced by `self.arm_to_expectation`). If the function was `predict_score`, then we could have a simpler implementation, or we wouldn't be bound by the expectations to match, just that the scores (and therefore arms) are ordered somehow.

To give an example, when we scored a context with the popularity policy before, it might've returned an item "a". The corresponding `predict_score` only needs to ensure that the score of "a" is highest. If you only had 2 items `a` and `b`, both `{a: 0.6, b:0.4}` and `{a:0.7, b:0.3}` would be valid results. In _expectation_, `b` still might get recommended more overall if `self.arm_to_expectations` had a higher value for `b`.

Reply:
Re: *distribution* of items - Yes, I did a few experiments to confirm that the *distribution* of items are consistent. As expected, Random and Epsilon Greedy match very closely. Popularity and Softmax approach the distribution yielded by the previous implementation for very large number of samples.

Reply:
Re: function name - I get what you are saying. To be fair this was already the case and I'm not sure if it warrants such a big change to the public API. Happy to chat more.